pkg/k8s/endpointsynchronizer/cep.go

```diff
@@ -225,48 +227,72 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
+
+					// If it fails it means the test from the previous patch failed
+					// so we can safely replace this node in the CNP status.
+					replaceCEPStatus := []k8s.JSONPatch{
+						{
+							OP:    "replace",
+							Path:  "/status",
+							Value: mdl,
+						},
+					}
+					var createStatusPatch []byte
+					createStatusPatch, err = json.Marshal(replaceCEPStatus)
+					if err != nil {
 						return err
+					}
+					_, err = ciliumClient.CiliumEndpoints(namespace).Patch(podName, types.JSONPatchType, createStatusPatch, "status")
+				default:
+					// We have no localCEP copy. We need to fetch it for updates, below.
+					// This is unexpected as there should be only 1 writer per CEP, this
+					// controller, and the localCEP created on startup will be used.
+					if localCEP == nil {
+						localCEP, err = ciliumClient.CiliumEndpoints(namespace).Get(podName, meta_v1.GetOptions{})
+						switch {
+						// The CEP doesn't exist in k8s. This is unexpetected but may occur
+						// if the endpoint was removed from k8s but not yet within the agent.
+						// Mark the CEP for creation on the next controller iteration. This
+						// may never occur if the controller is stopped on Endpoint delete.
+						case err != nil && k8serrors.IsNotFound(err):
+							needInit = true
+							return err
+
+						// We cannot read the upstream CEP. needInit will cause the next
+						// iteration to delete and create the CEP. This is an unexpected
+						// situation.
+						case err != nil && k8serrors.IsInvalid(err):
+							scopedLog.WithError(err).Warn("Invalid CEP during update")
+							needInit = true
+							return nil
+
+						// A real error
+						case err != nil:
+							scopedLog.WithError(err).Error("Cannot get CEP during update")
+							return err
+						}
+					}
 
-					// We cannot read the upstream CEP. needInit will cause the next
-					// iteration to delete and create the CEP. This is an unexpected
-					// situation.
-					case err != nil && k8serrors.IsInvalid(err):
-						scopedLog.WithError(err).Warn("Invalid CEP during update")
-						needInit = true
-						return nil
-
-					// A real error
-					case err != nil:
-						scopedLog.WithError(err).Error("Cannot get CEP during update")
+					// We have an object to reuse. Update and push it up. In the case of an
+					// update error, we retry in the next iteration of the controller using
+					// the copy returned by Update.
+					scopedLog.Debug("Updating CEP from local copy")
+					switch {
+					case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
+						localCEP, err = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(localCEP)
+						mdl.DeepCopyInto(&localCEP.Status)
+						_, err = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(localCEP)
+					default:
+						localCEP, err = ciliumClient.CiliumEndpoints(namespace).Update(localCEP)
+						mdl.DeepCopyInto(&localCEP.Status)
+						_, err = ciliumClient.CiliumEndpoints(namespace).Update(localCEP)
```

Comment:
I'm surprised that the value returned by Update and UpdateStatus isn't already correct. I'm not sure I understand why we do the second Update, or why we do mdl.DeepCopyInto after the first update.

Reply:
bad rebase. fixed